File: library-typescript/src/TemplateEnvironment.ts
import type { TemplateDelegate } from "handlebars";
import Handlebars from "handlebars";
import { BlockType } from "./BlockType.js";
import { sha256 } from "./hashing.js";
import type { TemplateCache } from "./TemplateCache.js";

/** Render-data key under which the active locale's translations are passed. */
export const CTX_TRANSLATIONS = "_translationsData";

type TranslationMap = Record<string, string>;

const ACTIVE_TRANSLATION_IDS = new Set<string>();

/**
 * A pair of Handlebars instances (one for HTML blocks, one for text blocks) plus
 * the helpers, compiling through a {@link TemplateCache}. Obtained from
 * {@link TemplateCache.environment} and shared by every template using that cache.
 */
export class TemplateEnvironment {
  private readonly cache: TemplateCache;
  private readonly htmlEnv: typeof Handlebars;
  private readonly textEnv: typeof Handlebars;

  constructor(cache: TemplateCache) {
    this.cache = cache;
    this.htmlEnv = this.buildEnv(BlockType.HTML);
    this.textEnv = this.buildEnv(BlockType.TEXT);
  }

  private envFor(type: BlockType): typeof Handlebars {
    return type === BlockType.TEXT ? this.textEnv : this.htmlEnv;
  }

  /**
   * Compiles a block, returning a cached result when the same source was
   * compiled before.
   *
   * @param type - The block's type, selecting the HTML or text instance.
   * @param source - The block's Handlebars source.
   * @returns The compiled template.
   */
  compile(type: BlockType, source: string): TemplateDelegate {
    return this.cachedCompile(this.envFor(type), `${type}:`, source, type === BlockType.TEXT);
  }

  private cachedCompile(
    env: typeof Handlebars,
    prefix: string,
    source: string,
    noEscape: boolean,
  ): TemplateDelegate {
    const key = prefix + sha256(source);
    const cached = this.cache.get(key);
    if (cached) return cached;
    const compiled = env.compile(source, { noEscape, ignoreStandalone: true });
    this.cache.put(key, source, compiled);
    return compiled;
  }

  private buildEnv(type: BlockType): typeof Handlebars {
    const env = Handlebars.create();
    const isText = type === BlockType.TEXT;
    const cachePrefix = `${type}:`;

    env.registerHelper("lookup", () => {
      throw new Error("The 'lookup' helper is disabled");
    });

    const self = this;

    env.registerHelper(
      "translate",
      function (this: Record<string, unknown>, _ctx: unknown, options: Handlebars.HelperOptions) {
        const defaultContent = options.fn(this);
        const translationId: string = options.hash.id ?? sha256(defaultContent);

        if (ACTIVE_TRANSLATION_IDS.has(translationId)) {
          throw new Error(
            `Circular translate reference detected: '${translationId}' is already being rendered`,
          );
        }
        ACTIVE_TRANSLATION_IDS.add(translationId);
        try {
          const localeMap = (options.data?.root as Record<string, unknown>)?.[
            CTX_TRANSLATIONS
          ] as TranslationMap | null;
          const content = resolveTranslation(localeMap, translationId, defaultContent);
          const tpl = self.cachedCompile(env, cachePrefix, content, isText);
          const result = tpl(options.data?.root as object);
          return new Handlebars.SafeString(result);
        } finally {
          ACTIVE_TRANSLATION_IDS.delete(translationId);
        }
      },
    );

    return env;
  }
}

function resolveTranslation(
  localeMap: TranslationMap | null,
  key: string,
  defaultContent: string,
): string {
  if (!localeMap) return defaultContent;
  return key in localeMap ? localeMap[key] : defaultContent;
}

File: library-java/lib/src/main/java/eu/postyard/templates/TemplateEnvironment.java
package eu.postyard.templates;

import com.github.jknack.handlebars.EscapingStrategy;
import com.github.jknack.handlebars.Handlebars;
import java.io.IOException;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * The Handlebars setup shared by every {@link Template} that uses a given {@link TemplateCache}:
 * one {@link Handlebars} instance for HTML blocks and one for text blocks, the custom helpers, and
 * the compilation that reads from and writes to the cache. Holding this here means the Handlebars
 * instances and helpers are built once per cache rather than once per template.
 *
 * <p>The HTML instance escapes interpolated values; the text instance uses {@link
 * EscapingStrategy#NOOP} so values pass through verbatim. Two helpers are registered:
 *
 * <ul>
 *   <li>{@code lookup} is disabled and throws if used.
 *   <li>{@code translate} resolves a localized string by id from the locale supplied in the render
 *       context, falling back to the block's inlined text, then compiles and renders that string so
 *       a translation may itself contain Handlebars expressions. Re-entering the same translation
 *       id while it is already being rendered is rejected as a cycle.
 * </ul>
 */
final class TemplateEnvironment {

  /** Key under which the active locale's translations are passed in the render context. */
  static final String CTX_TRANSLATIONS = "_translationsData";

  private static final ThreadLocal<Set<String>> ACTIVE_TRANSLATION_IDS =
      ThreadLocal.withInitial(HashSet::new);

  private final TemplateCache cache;
  private final Handlebars htmlHandlebars;
  private final Handlebars textHandlebars;

  TemplateEnvironment(TemplateCache cache) {
    this.cache = cache;
    this.htmlHandlebars = buildHandlebars(BlockType.HTML);
    this.textHandlebars = buildHandlebars(BlockType.TEXT);
  }

  /**
   * Compiles a block of the given type, returning a cached result when the same source has been
   * compiled before.
   *
   * @param type the block's type, which selects the HTML or text instance
   * @param source the block's Handlebars source
   * @return the compiled template
   * @throws IOException if the source cannot be compiled
   */
  com.github.jknack.handlebars.Template compile(BlockType type, String source) throws IOException {
    Handlebars hb = type == BlockType.HTML ? htmlHandlebars : textHandlebars;
    return cachedCompile(hb, type.name().toLowerCase() + ":", source);
  }

  private Handlebars buildHandlebars(BlockType type) {
    Handlebars hb = new Handlebars();
    if (type == BlockType.TEXT) {
      hb.with(EscapingStrategy.NOOP);
    }
    String cachePrefix = type.name().toLowerCase() + ":";

    hb.registerHelper(
        "lookup",
        (Object ctx, com.github.jknack.handlebars.Options opts) -> {
          throw new UnsupportedOperationException("The 'lookup' helper is disabled");
        });

    hb.registerHelper(
        "translate",
        (Object value, com.github.jknack.handlebars.Options opts) -> {
          String defaultContent = opts.fn(value).toString();
          String translationId = opts.hash("id");
          if (translationId == null) {
            translationId = Hashing.sha256(defaultContent);
          }

          Set<String> active = ACTIVE_TRANSLATION_IDS.get();
          if (!active.add(translationId)) {
            throw new IllegalStateException(
                "Circular translate reference detected: '"
                    + translationId
                    + "' is already being rendered");
          }
          try {
            @SuppressWarnings("unchecked")
            Map<String, String> localeMap = (Map<String, String>) opts.data(CTX_TRANSLATIONS);
            String content = resolveTranslation(localeMap, translationId, defaultContent);
            return new Handlebars.SafeString(
                cachedCompile(hb, cachePrefix, content).apply(opts.context));
          } finally {
            active.remove(translationId);
          }
        });

    return hb;
  }

  private com.github.jknack.handlebars.Template cachedCompile(
      Handlebars hb, String cachePrefix, String content) throws IOException {
    String hash = cachePrefix + Hashing.sha256(content);
    com.github.jknack.handlebars.Template cached = cache.get(hash);
    if (cached != null) return cached;
    com.github.jknack.handlebars.Template compiled = hb.compileInline(content);
    cache.put(hash, content, compiled);
    return compiled;
  }

  private static String resolveTranslation(
      Map<String, String> localeMap, String key, String defaultContent) {
    if (localeMap == null) return defaultContent;
    return localeMap.getOrDefault(key, defaultContent);
  }
}
